### ISSUE TITLE
SASL mechanism SCRAM-SHA-256 not working

### ISSUE THREAD

**Issue by vinodres** [confluentinc/confluent-kafka-dotnet#169]
I am trying to connect to Kafka brokers v0.10.2 running on Linux CentOS 7, where recently security was enabled. 

I am using Confluent.Kafka v0.9.5 and my test producer and consumer applications are running on Windows 7 with .NET 4.6.1 framework.

I am trying to use the below configuration for the producer.
```
var config = new Dictionary<string, object>
            {
                { "bootstrap.servers", brokerList },
                { "client.id", $"MyCorp.Poc.Producer.1" },
                { "sasl.mechanism", "SCRAM-SHA-256" },
                { "security.protocol", "SASL_PLAINTEXT" },
                { "sasl.jaas.config", @"org.apache.kafka.common.security.scram.ScramLoginModule required username=""testuser"" password=""test$1245"";" },
                { "default.topic.config", new Dictionary<string, object>()
                    {
                        { "acks", -1 },
                        { "request.timeout.ms", 5000 },

                    }
                }
            };
```
I am getting following exception
```
Unhandled Exception: System.InvalidOperationException: No such configuration property: "sasl.mechanism"
   at Confluent.Kafka.Impl.SafeConfigHandle.Set(String name, String value)
   at Confluent.Kafka.Producer.<>c__DisplayClass22_0.<.ctor>b__2(KeyValuePair`2 kvp)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at Confluent.Kafka.Producer..ctor(IEnumerable`1 config, Boolean manualPoll, Boolean disableDeliveryReports)
   at Confluent.Kafka.Producer`2..ctor(IEnumerable`1 config, ISerializer`1 keySerializer, ISerializer`1 valueSerializer, Boolean manualPoll, Boolean disableDeliveryReports)
   at Confluent.Kafka.Producer`2..ctor(IEnumerable`1 config, ISerializer`1 keySerializer, ISerializer`1 valueSerializer)
```
Is SASL PLAINTEXT security protocol supported with Confluent.Kafka .net client? If not, are there any other options?

We have clients with both JAVA and ,NET, which security mechanism is going to work for both type of clients?

**Comment by treziac** [comment]
jaas is java only.
I'm not well known in SASL - you can check on librdkafka for more details, there were many posts about this

SASL PLAINTEXT is supported, you can see here for more details:
https://github.com/edenhill/librdkafka/wiki/Using-SASL-with-librdkafka

**Comment by edenhill** [comment]
The configuration property is `sasl.mechanisms` (plural, even though you can only specify one).

The next librdkafka release will support SASL-SCRAM, but v0.9.5 does not.
If you want an early start you can try librdkafka master.

**Comment by vinodres** [comment]
Thank you eden. I will fix the typo and will give a try with master branch.

**Comment by kitron** [comment]
Eden,
Do you have any timeframe for next librdkafka release that supports SASL-SCRAM?

Thanks

**Comment by edenhill** [comment]
@kitron The next feature release is due some time in june

**Comment by kitron** [comment]
Thank you Eden!

**Comment by scovel** [comment]
Is there a branch with the SASL implementation that I could pull and test?  I need to interop with the Java guys in my shop.

Thanks

**Comment by treziac** [comment]
SASL (GSSAPI and PLAIN) is supported on the current release - check https://github.com/edenhill/librdkafka/wiki/Using-SASL-with-librdkafka for usage. There is no example for now on this repo (you will find some on issues though)

SCRAM-SHA-256, SCRAM-SHA-512 are implemented on librdkafka master - you will have to build and link your own librdkafka dll in place of librdkafka.redist for this to work

**Comment by kavyashivakumar** [comment]
@vinodres hi, were you able to connect successfully to kafka using sasl on .net? I've very similar envt as yours and i'm facing quite few problems. (have posted them in issues too but waiting on response)

would you be able to post sample dictionary config all you've to send ? it would be extremely helpful. desperately waiting for some help. (i'm using GSSAPI )

**Comment by vinodres** [comment]
I am waiting for next release as v0.9.5 does not support SASL(**SCRAM** and PLAIN).

**Comment by vinodres** [comment]
@edenhill 
Hi Eden, 
I was finally able to spend some time to perform SASL-SCRAM testing.

I build the librdkafka master branch and tried to test SASL-SCRAM and PLAINTEXT security protocol.

The test producer is a .net console application running on Windows 7 and in 32bit (x86) mode.

The Library version now appears as **0.9.5-pre0** when I execute Confluent.Kafka.Library.VersionString

But when I tried to connect to Kafka broker, Now I am getting following error

> 2017-06-08 162041.PM Error: sasl_plaintext://10.56.178.106:9092/bootstrap: Failed to initialize SASL authentication: SASL Handshake not supported by broker (required by mechanism SCRAM-SHA-256): try api.version.request=true

Just wondering if I am doing something wrong. Here is how my producer config looks like
 ```
          var config = new Dictionary<string, object>
            {
                {"bootstrap.servers", "10.56.178.106:9092,10.56.178.108:9092,10.56.178.112:9092"},
                {"client.id", $"MyCorp.Poc.Producer"},
                {"sasl.mechanisms", "SCRAM-SHA-256"},
                {"security.protocol", "SASL_PLAINTEXT"},
                {"sasl.username", "testuser"},
                {"sasl.password", "test1234"},
                {
                    "default.topic.config", new Dictionary<string, object>()
                    {
                        {"acks", -1},
                        {"request.timeout.ms", 5000},
                    }
                }
            };
```

This is how the JAVA teams have constructed their config and they are able to publish messages

```
    Properties props = new Properties();
    props.put("bootstrap.servers", "10.56.178.106:9092,10.56.178.108:9092,10.56.178.112:9092");
    props.put("acks", "1");
    props.put("sasl.mechanism", "SCRAM-SHA-256");
    props.put("security.protocol", "SASL_PLAINTEXT");
    props.put("sasl.jaas.config", "org.apache.kafka.common.security.scram.ScramLoginModule required username=\"testuser\" password=\"test1234\";");
    props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
    props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
```

- Vinod

**Comment by treziac** [comment]
Sorry, a bit late for answer, but in case you didn't solve it, as stated in error message you have to provide "api.version.request" to true in config.
This will be set to true by default in a future release as more and more people are using kafka 0.10+

**Comment by vinodres** [comment]
Thank you @treziac , that did the trick. It is now working.

**Comment by vinodres** [comment]
After my previous successful test with Confluent.Kafka 0.9.5 and librakafka master branch, I noticed that Confluent.Kafka 0.11.0-RC1 was released. I then upgraded my test application to it and now I am unable to produce or consume. Just wondering if I have to change anything on producer and/or consumer config? The topic was created with 3 partitions.

Here is the producer config
```
var config = new Dictionary<string, object>
            {
                {"bootstrap.servers", "10.56.178.106:9092,10.56.178.108:9092,10.56.178.112:9092"},
                {"client.id", "Vinod_Test_Producer_Client_1"},
                {"sasl.mechanisms", "SCRAM-SHA-256"},
                {"security.protocol", "SASL_PLAINTEXT"},
                {"sasl.username", "testuser"},
                {"sasl.password", "test1234"},
                { "api.version.request", "true" },
                { "debug", "cgrp,topic,fetch"},
                {
                    "default.topic.config", new Dictionary<string, object>()
                    {
                        {"acks", -1},
                        {"request.timeout.ms", 5000},
                    }
                }
            };
```
Here are the logs from producer console app. 

> 0.11.0-RC1
> 
> -----------------------------------------------------------------------
> Producer Vinod_Test_Producer_Client_1#producer-1 producing on topic MyCorp.Messaging.PocRC11.Topic.
> -----------------------------------------------------------------------
> To create a kafka message with UTF-8 encoded key/value message:
> > key value<Enter>
> To create a kafka message with empty key and UTF-8 encoded value:
> > value<enter>
> Ctrl-C to quit.
> 
> > HelloRC11_Secure_1
> 2017-06-28 161049.PM Log: [thrd:app]: New local topic: MyCorp.Messaging.PocRC11.Topic
> 2017-06-28 161049.PM Log: [thrd:app]: NEW MyCorp.Messaging.PocRC11.Topic [-1] 00724A50 (at rd_kafka_topic_new0:282)
> 2017-06-28 161050.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161051.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161052.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161053.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161054.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161055.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161056.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161057.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161058.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161059.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161100.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161101.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161102.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161103.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161104.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161105.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata
> 2017-06-28 161106.PM Log: [thrd:main]: Topic MyCorp.Messaging.PocRC11.Topic partition count is zero: should refresh metadata

Here is how my consumer config looks like
```
new Dictionary<string, object>
            {
                { "group.id", "Vinod_Test_Client.1.Group.1"},
                { "client.id", "Vinod_Test_Client.1" },
                { "enable.auto.commit", false },
                { "auto.commit.interval.ms", 5000 },
                { "bootstrap.servers", brokerList },
                {"sasl.mechanisms", "SCRAM-SHA-256"},
                {"security.protocol", "SASL_PLAINTEXT"},
                {"sasl.username", "kafkauser"},
                {"sasl.password", "Espn1234"},
                { "api.version.request", "true" },
                { "debug", "cgrp,topic,fetch"},
                { "default.topic.config", new Dictionary<string, object>()
                    {
                        { "auto.offset.reset", "latest" }
                    }
                }
            };
```
Here are the logs from consumer console app

> 0.11.0-RC1
> 7|2017-06-28 16:10:22.172|Vinod_Test_Client.1#consumer-1|MEMBERID| [thrd:app]: Group "Vinod_Test_Client.1.Group.1": updating member id "(not-set)" -> ""
> 7|2017-06-28 16:10:22.210|Vinod_Test_Client.1#consumer-1|BRKREASSIGN| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" management reassigned from broker (none) to :0/internal
> 7|2017-06-28 16:10:22.211|Vinod_Test_Client.1#consumer-1|CGRPSTATE| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" changed state init -> wait-broker (v1, join-state init)
> 7|2017-06-28 16:10:22.211|Vinod_Test_Client.1#consumer-1|BRKASSIGN| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" management assigned to broker :0/internal
> 7|2017-06-28 16:10:22.222|Vinod_Test_Client.1#consumer-1|CGRPOP| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" received op SUBSCRIBE (v0) in state wait-broker (join state init, v1 vs 0)
> 7|2017-06-28 16:10:22.222|Vinod_Test_Client.1#consumer-1|SUBSCRIBE| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": subscribe to new subscription of 1 topics (join state init)
> 7|2017-06-28 16:10:22.222|Vinod_Test_Client.1#consumer-1|UNSUBSCRIBE| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": unsubscribe from current unset subscription of 0 topics (leave group=no, join st
> ate init, v1)
> 7|2017-06-28 16:10:22.223|Vinod_Test_Client.1#consumer-1|CGRPJOINSTATE| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" changed join state init -> wait-unassign (v1, state wait-broker)
> 7|2017-06-28 16:10:22.223|Vinod_Test_Client.1#consumer-1|UNASSIGN| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": unassign done in state wait-broker (join state wait-unassign): without new assignme
> nt: unassign (no previous assignment)
> 7|2017-06-28 16:10:22.223|Vinod_Test_Client.1#consumer-1|CGRPJOINSTATE| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" changed join state wait-unassign -> init (v1, state wait-broker)
> 7|2017-06-28 16:10:22.224|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:22.227|Vinod_Test_Client.1#consumer-1|CGRPOP| [thrd:main]: Group "Vinod_Test_Client.1.Group.1" received op GET_SUBSCRIPTION (v0) in state wait-broker (join state init, v1 vs 0)
> 2017-06-28 161022.PM Subscribed to: [MyCorp.Messaging.PocRC11.Topic]
> Started consumer, press enter to stop consuming
> 7|2017-06-28 16:10:23.223|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:24.237|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:25.251|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:26.266|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:27.280|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker
> 7|2017-06-28 16:10:28.294|Vinod_Test_Client.1#consumer-1|CGRPQUERY| [thrd:main]: Group "Vinod_Test_Client.1.Group.1": no broker available for coordinator query: intervaled in state wait-broker

**Comment by treziac** [comment]
There was a bug on SASL on RC1, it will be corrected on RC2 (if you want, you can try the latest artefacts from librdkafka appveyor, but they aren't package for now)

**Comment by vinodres** [comment]
I will wait for RC2. For now I will revert back to 0.9.5 with librdkafka master branch. Thanks for a quick response.

**Comment by treziac** [comment]
If you want, you can also use confluent.kafka 0.11.0-RC1 with librdkafka master branch (or last artefacts from 0.11.x which should work but only tested master on my side : https://ci.appveyor.com/project/edenhill/librdkafka/build/0.9.4-R-post139)

Last .NET package has not been throughfully tested yet, if you use it don't mind report any feedback ;)

**Comment by vinodres** [comment]
Just closing the loop on this issue. Authentication is now working after upgrading to Confluent.Kafka 0.11.0.

Here is the config I use for Producer
var config = new Dictionary<string, object>()
                {
                    {"bootstrap.servers", "mycorpmessagebroker01:9092,mycorpmessagebroker02:9092,mycorpmessagebroker03:9092"},
                    {"client.id", "Vinod_Test_Client"},
                    //begin security settings
                    {"sasl.mechanisms", "SCRAM-SHA-256"},
                    {"security.protocol", "SASL_PLAINTEXT"},
                    {"sasl.username", "testuser"},
                    {"sasl.password", "testpwd"},
                    {"api.version.request", "true"}
                    //end security settings
                    {
                        "default.topic.config", new Dictionary<string, object>()
                        {
                            {"acks", -1}
                        }
                    }
                };

I opened another issue #285 . After implementing authentication, ran into a CPU issue when I am using more number of consumers per topic.
